pkg/ha/filter.go

```diff
@@ -49,42 +51,136 @@ func (h *Filter) Process(_ *http.Request, wr *prompb.WriteRequest) error {
+		// This case covers the instance when we have backfill and data in current lease to ingest.
+		// The data has already been filtered out so there is nothing to do.
+	}
+	return nil
+}
+
+func toPromModelTime(t time.Time) int64 {
+	return int64(model.TimeFromUnixNano(t.UnixNano()))
+}
+
+func (h *Filter) filterBackfill(wr *prompb.WriteRequest, minTIncl, maxTExcl time.Time, cluster, replica string) (bool, error) {
+	hasBackfill := false
+	backfillStart := minTIncl
+	for {
+		if !backfillStart.Before(maxTExcl) {
+			break
+		}
+
+		keepRangeStart, keepRangeEnd, err := h.service.GetBackfillLeaseRange(backfillStart, maxTExcl, cluster, replica)
+		if err != nil {
+			if err == ErrNoLeasesInRange {
+				// Nothing else to keep.
+				break
+			}
+			return false, fmt.Errorf("could not check backfill ha lease: %#v", err)
+		}
+
+		hasBackfill = true
+		// Filter out all samples before the keep range.
+		if !keepRangeStart.Before(backfillStart) {
+			filterOutSampleRange(
+				wr,
+				toPromModelTime(backfillStart),
+				toPromModelTime(keepRangeStart),
+			)
 		}
+
+		backfillStart = keepRangeEnd
+	}
+
+	// If we had backfill, we need to filter out the range after the keep range
+	// till the start of the lease (end of possible backfill).
+	if hasBackfill {
+		filterOutSampleRange(
+			wr,
+			toPromModelTime(backfillStart),
+			toPromModelTime(maxTExcl),
+		)
+	}
+
+	return hasBackfill, nil
+}
+
+func filterOutSampleRange(wr *prompb.WriteRequest, timeStartIncl, timeEndExcl int64) {
+	for i := range wr.Timeseries {
+		t := &wr.Timeseries[i]
+		numAccepted := 0
+		for j := range t.Samples {
+			sample := t.Samples[j]
+			if sample.Timestamp >= timeStartIncl && sample.Timestamp < timeEndExcl {
+				continue
+			}
+			t.Samples[numAccepted] = sample
+			numAccepted++
+		}
+		for j := numAccepted; j < len(t.Samples); j++ {
+			t.Samples[j] = prompb.Sample{}
+		}
+		t.Samples = t.Samples[:numAccepted]
+	}
+}
+
+// finalFiltering goes through all the `Timeseries` of a `WriteRequest` filtering
+// out any instances without any samples. If the timeseries does contain samples,
+// it filters out the HA replica labels so it won't create different series
+// based on that label value.
+func finalFiltering(wr *prompb.WriteRequest) {
+	numAccepted := 0
+	for i := range wr.Timeseries {
+		t := &wr.Timeseries[i]
 		if len(t.Samples) == 0 {
-			// Empty the labels for posterity.
-			t.Labels = make([]prompb.Label, 0)
 			continue
 		}
-
+		wr.Timeseries[numAccepted] = *t
+		numAccepted++
 		// Drop __replica__ labelSet from samples,
 		// we don't want samples from the same Prometheus
 		// HA set to become different series.
 		for ind, value := range t.Labels {
-			if value.Name == ReplicaNameLabel && value.Value == replicaName {
+			if value.Name == ReplicaNameLabel {
 				t.Labels = append(t.Labels[:ind], t.Labels[ind+1:]...)
 				break
 			}
 		}
 	}
-
-	return nil
+	wr.Timeseries = wr.Timeseries[:numAccepted]
+	for j := numAccepted; j < len(wr.Timeseries); j++ {
```

Comment:
this loop has to come before the previous line `wr.Timeseries = wr.Timeseries[:numAccepted]` because that line sets len(wr.Timeseries)